lib/common/ipc_client.c

```diff
@@ -305,40 +305,67 @@ call_api_dispatch(pcmk_ipc_api_t *api, xmlNode *message)
     return false;
 }
 
-/*!
- * \internal
- * \brief Dispatch data read from IPC source
- *
- * \param[in] buffer     Data read from IPC
- * \param[in] length     Number of bytes of data in buffer (ignored)
- * \param[in] user_data  IPC object
- *
- * \return Always 0 (meaning connection is still required)
+#define MORE_MESSAGES -1
+
+/* Do the hard work of dispatch_ipc_data.  This is split out into its own
+ * function so it can be shared with pcmk__send_ipc_request.  The return
+ * value is as follows:
  *
- * \note This function can be used as a main loop IPC dispatch callback.
+ * pcmk_rc_error: Some error occurred.  It's up to the caller to decide
+ *                what to do with that fact.
+ * pcmk_rc_ok: There are no more messages expected from the server.  Quit
+ *             reading.
+ * MORE_MESSAGES: There are more messages expected from the server.  Keep
+ *                reading.
  */
 static int
-dispatch_ipc_data(const char *buffer, ssize_t length, gpointer user_data)
+dispatch_ipc_data(const char *buffer, pcmk_ipc_api_t *api)
 {
-    pcmk_ipc_api_t *api = user_data;
+    bool more = false;
     xmlNode *msg;
 
-    CRM_CHECK(api != NULL, return 0);
-
     if (buffer == NULL) {
         crm_warn("Empty message received from %s IPC",
                  pcmk_ipc_name(api, true));
-        return 0;
+        return pcmk_rc_error;
     }
 
     msg = string2xml(buffer);
     if (msg == NULL) {
         crm_warn("Malformed message received from %s IPC",
                  pcmk_ipc_name(api, true));
-        return 0;
+        return pcmk_rc_error;
```

Comment:
EPROTO is the easiest for anything unrecognized

Reply:
Fixed.